### PR TITLE
Remove undocumented certificate APIs

### PR DESCRIPTION
**Pull request by TravisEz13**
Addresses #3792 

The code paths deleted where  using the undocumented API on WIn8 and newer for the following purposes:
* Faster filtering
* Getting two properties (left empty on Win7 and below)
* Logging using Certificate components when a cert is deleted or copied.

After the change, all the code uses public APIs.  Filtering is done in PowerShell using existing code.  I don't believe the logging is needed.

### CODE REVIEW

File: src/Microsoft.PowerShell.Security/security/CertificateProvider.cs [hunk 1]
```diff
@@ -3598,81 +3443,25 @@ public List<DnsNameRepresentation> DnsNameList
...
-
-                        hr = Security.NativeMethods.CCGetCertNameList(
-                                    certHandle,
-                                    nameType,
-                                    Security.NativeMethods.CryptDecodeFlags.CRYPT_DECODE_ENABLE_IA5CONVERSION_FLAG,
-                                    out cUnicodeName,
-                                    out papwszUnicodeNames);
-                        if (hr != Security.NativeConstants.S_OK)
+                        // Get the part after 'DNS Name='
+                        if(nameLine.Length > 9)
```
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.Security/security/CertificateProvider.cs]: Any validation needed here?  Prefer to have const DnsNamePrefixLength used than 9

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]
Is there sufficient test coverage?

File: src/Microsoft.PowerShell.Security/security/CertificateProvider.cs [hunk 2]
```diff
@@ -2531,14 +2448,7 @@ protected override object GetItemDynamicParameters(string path)
...
         {
-            if (DownLevelHelper.NativeFilteringSupported())
-            {
-                return new CertificateProviderDynamicParameters();
-            }
-            else
-            {
-                return new CertificateProviderCodeSigningDynamicParameters();
-            }
+            return new CertificateProviderCodeSigningDynamicParameters();
```
Comment (TravisEz13) [inline src/Microsoft.PowerShell.Security/security/CertificateProvider.cs]: This is the code path that is always taken on `Windows PowerShell`.  It should work.

File: src/Microsoft.PowerShell.Security/security/CertificateProvider.cs [hunk 3]
```diff
@@ -2092,10 +2030,7 @@ private void DoRemove(X509Certificate2 cert, bool fDeleteKey, bool fMachine, str
                     CommitUserDS(context.hCertStore);
                 }
 
-                if (DownLevelHelper.LogCertChangeSupported())
-                {
-                    Security.NativeMethods.LogCertDelete(fMachine, cert.Handle);
-                }
+                //TODO: Log Cert Delete
```
Comment (TravisEz13) [inline src/Microsoft.PowerShell.Security/security/CertificateProvider.cs]: I'm not sure we need these, but I'll open an issue or remove them based on my investigation.

File: src/Microsoft.PowerShell.Security/security/CertificateProvider.cs (hunk 1 above)
Comment (TravisEz13) [inline src/Microsoft.PowerShell.Security/security/CertificateProvider.cs]: Yes, I was planning on adding it with the support for punyCode.  The only way I can see puny code being supported is a different record format.

File: src/Microsoft.PowerShell.Security/security/CertificateProvider.cs [hunk 4]
```diff
@@ -2173,18 +2108,7 @@ private void DoMove(string destination, X509Certificate2 cert, X509NativeStore s
...
-
-                    if (cert.HasPrivateKey &&
-                        String.Equals(store.StoreName, "MY", StringComparison.OrdinalIgnoreCase))
-                    {
-                        Security.NativeMethods.LogCertCopy(fMachine, cert.Handle);
-                    }
-
-                    Security.NativeMethods.LogCertDelete(fMachine, cert.Handle);
-                }
+                //TODO: log cert move
```
Comment (TravisEz13) [inline src/Microsoft.PowerShell.Security/security/CertificateProvider.cs]: filed #3826, until this is resolved

File: src/Microsoft.PowerShell.Security/security/CertificateProvider.cs (hunk 3 above)
Comment (TravisEz13) [inline src/Microsoft.PowerShell.Security/security/CertificateProvider.cs]: filed #3826, until this is resolved

**Review by SteveL-MSFT — COMMENTED** [review]
My concerns are addressed, but would like @LeeHolmes to review before we merge

File: test/powershell/Modules/Microsoft.PowerShell.Security/CertificateProvider.Tests.ps1 [hunk 5]
```diff
@@ -0,0 +1,138 @@
+Import-Module (Join-Path -Path $PSScriptRoot 'certificateCommon.psm1') -Force
```
Comment (PaulHigin) [inline test/powershell/Modules/Microsoft.PowerShell.Security/CertificateProvider.Tests.ps1]: Please put this test prep inside "if ($IsWindows) {}"

**Review by PaulHigin — CHANGES_REQUESTED** [review]

File: src/Microsoft.PowerShell.Security/security/CertificateProvider.cs [hunk 6]
```diff
@@ -508,19 +457,9 @@ public IntPtr GetNextCert(IntPtr certContext)
             }
             if (Valid)
             {
-                if (_filterHandle != null)
```
Comment (TravisEz13) [inline src/Microsoft.PowerShell.Security/security/CertificateProvider.cs]: probably just performance.  Looking at the CmdLet we only supported filtering by codesigning certs which the tests pass after the change.

File: src/Microsoft.PowerShell.Security/security/CertificateProvider.cs [hunk 7]
```diff
@@ -3541,32 +3403,15 @@ public List<EnhancedKeyUsageRepresentation> EnhancedKeyUsageList
         /// </summary>
         public EnhancedKeyUsageProperty(X509Certificate2 cert)
         {
-            if (DownLevelHelper.NativeFilteringSupported())
+            foreach (X509Extension extension in cert.Extensions)
             {
-                Collection<string> ekuCollection = System.Management.Automation.Internal.SecuritySupport.GetCertEKU(cert);
-
-                foreach (string oidString in ekuCollection)
+                if (extension.Oid.FriendlyName == "Enhanced Key Usage")
```
Comment (TravisEz13) [inline src/Microsoft.PowerShell.Security/security/CertificateProvider.cs]: Comparing the value would be better.  I'll update the code.

File: src/Microsoft.PowerShell.Security/security/CertificateProvider.cs [hunk 8]
```diff
@@ -3598,81 +3446,65 @@ public List<DnsNameRepresentation> DnsNameList
...
+                            catch(System.ArgumentException)
+                            {
+                                // The name is not valid punyCode, assume it's valid ascii.
+                                unicodeName = name;
+                            }
+
+                            dnsName = new DnsNameRepresentation(name,unicodeName);
+
+                            // Only add the name if it is not the same as an existing name.
+                            if(!_dnsList.Contains(dnsName))
```
Comment (TravisEz13) [inline src/Microsoft.PowerShell.Security/security/CertificateProvider.cs]: The old code gave all variations back in the results.  So, no.

File: test/powershell/Modules/Microsoft.PowerShell.Security/CertificateProvider.Tests.ps1 (hunk 5 above)
Comment (TravisEz13) [inline test/powershell/Modules/Microsoft.PowerShell.Security/CertificateProvider.Tests.ps1]: What is the concern here?

File: src/Microsoft.PowerShell.Security/security/CertificateProvider.cs (hunk 6 above)
Comment (TravisEz13) [inline src/Microsoft.PowerShell.Security/security/CertificateProvider.cs]: line 2636, is filtering.

File: src/Microsoft.PowerShell.Security/security/CertificateProvider.cs [hunk 9]
```diff
@@ -2531,14 +2448,7 @@ protected override object GetItemDynamicParameters(string path)
         /// </returns>
         protected override object GetChildItemsDynamicParameters(string path, bool recurse)
         {
-            if (DownLevelHelper.NativeFilteringSupported())
-            {
-                return new CertificateProviderDynamicParameters();
```
Comment (TravisEz13) [inline src/Microsoft.PowerShell.Security/security/CertificateProvider.cs]: #resolved

File: test/powershell/Modules/Microsoft.PowerShell.Security/CertificateProvider.Tests.ps1 (hunk 5 above)
Comment (PaulHigin) [inline test/powershell/Modules/Microsoft.PowerShell.Security/CertificateProvider.Tests.ps1]: It seems like we can skip the prep work if not running on Windows.

File: src/Microsoft.PowerShell.Security/security/CertificateProvider.cs (hunk 7 above)
Comment (TravisEz13) [inline src/Microsoft.PowerShell.Security/security/CertificateProvider.cs]: Except that is the OID for a specific EKU, not specifying an EKU.  The OID should be 2.5.29.37

Comment (TravisEz13) [inline src/Microsoft.PowerShell.Security/security/CertificateProvider.cs]: #Resolved

File: src/Microsoft.PowerShell.Security/security/CertificateProvider.cs [hunk 10]
```diff
@@ -3541,32 +3403,15 @@ public List<EnhancedKeyUsageRepresentation> EnhancedKeyUsageList
...
-            if (DownLevelHelper.NativeFilteringSupported())
+            foreach (X509Extension extension in cert.Extensions)
             {
-                Collection<string> ekuCollection = System.Management.Automation.Internal.SecuritySupport.GetCertEKU(cert);
-
-                foreach (string oidString in ekuCollection)
+                if (extension.Oid.FriendlyName == "Enhanced Key Usage")
                 {
-                    if (!String.IsNullOrEmpty(oidString))
+                    X509EnhancedKeyUsageExtension ext = (X509EnhancedKeyUsageExtension)extension;
```
Comment (TravisEz13) [inline src/Microsoft.PowerShell.Security/security/CertificateProvider.cs]: #Resolved

File: src/Microsoft.PowerShell.Security/security/CertificateProvider.cs [hunk 11]
```diff
@@ -3581,6 +3426,9 @@ public EnhancedKeyUsageProperty(X509Certificate2 cert)
     public sealed class DnsNameProperty
     {
         private List<DnsNameRepresentation> _dnsList = new List<DnsNameRepresentation>();
+        private System.Globalization.IdnMapping idnMapping = new System.Globalization.IdnMapping();
+        private static string dnsNamePrefix = "DNS Name=";
```
Comment (TravisEz13) [inline src/Microsoft.PowerShell.Security/security/CertificateProvider.cs]: #resolved

File: src/Microsoft.PowerShell.Security/security/CertificateProvider.cs [hunk 12]
```diff
@@ -3598,81 +3446,65 @@ public List<DnsNameRepresentation> DnsNameList
...
-                                    Security.NativeMethods.CryptDecodeFlags.CRYPT_DECODE_ENABLE_IA5CONVERSION_FLAG,
-                                    out cUnicodeName,
-                                    out papwszUnicodeNames);
-                        if (hr != Security.NativeConstants.S_OK)
-                        {
-                            throw Marshal.GetExceptionForHR(hr);
-                        }
-                        if (cPunycodeName != cUnicodeName)
+                        // Get the part after 'DNS Name='
+                        if(nameLine.Length > dnsNamePrefix.Length && nameLine.StartsWith(dnsNamePrefix, System.StringComparison.InvariantCulture))
```
Comment (TravisEz13) [inline src/Microsoft.PowerShell.Security/security/CertificateProvider.cs]: #resolved

File: src/Microsoft.PowerShell.Security/security/CertificateProvider.cs [hunk 13]
```diff
@@ -3598,81 +3446,65 @@ public List<DnsNameRepresentation> DnsNameList
...
-            if (DownLevelHelper.NativeFilteringSupported())
+            string name;
+            string unicodeName;
+            DnsNameRepresentation dnsName;
+            _dnsList = new List<DnsNameRepresentation>();
+
+            // extract DNS name from subject distinguish name
+            // if it exists and does not contain a comma
+            // a comma, indicates it is not a DNS name
+            if(cert.Subject.Length > distinguishedNamePrefix.Length &&
```
Comment (TravisEz13) [inline src/Microsoft.PowerShell.Security/security/CertificateProvider.cs]: #resolved

File: src/Microsoft.PowerShell.Security/security/CertificateProvider.cs (hunk 8 above)
Comment (TravisEz13) [inline src/Microsoft.PowerShell.Security/security/CertificateProvider.cs]: #resolved

File: test/powershell/Modules/Microsoft.PowerShell.Security/CertificateProvider.Tests.ps1 (hunk 5 above)
Comment (TravisEz13) [inline test/powershell/Modules/Microsoft.PowerShell.Security/CertificateProvider.Tests.ps1]: The only imports functions, the actual prep work is done in the `beforeall` of the feature tests and is skipped on non-windows platforms.

**Review by PaulHigin — APPROVED** [review]
LGTM

**Review by SteveL-MSFT — APPROVED** [review]

**Pull request merged**